auth/src/main/java/com/firebase/ui/auth/provider/FacebookProvider.java

```diff
@@ -184,6 +163,44 @@ public void onError(FacebookException error) {
         Bundle extra = new Bundle();
         extra.putString(ERROR, "error");
         extra.putString(ERROR_MSG, error.getMessage());
-        mCallbackObject.onFailure(extra);
+        onFailure(extra);
+    }
+
+    private IdpResponse createIdpResponse(String email, LoginResult loginResult) {
+        return new IdpResponse(
+                FacebookAuthProvider.PROVIDER_ID,
+                email,
+                loginResult.getAccessToken().getToken());
+    }
+
+    private void onSuccess(String email, LoginResult loginResult) {
+        gcCallbackManager();
+        mCallbackObject.onSuccess(createIdpResponse(email, loginResult));
+    }
+
+    private void onFailure(Bundle bundle) {
+        gcCallbackManager();
+        mCallbackObject.onFailure(bundle);
+    }
+
+    private void gcCallbackManager() {
+        // sCallbackManager must be static to prevent it from being destroyed if the activity
+        // containing FacebookProvider dies.
+        // In startLogin(Activity), LoginManager#registerCallback(CallbackManager, FacebookCallback)
+        // stores the FacebookCallback parameter--in this case a FacebookProvider instance--into
+        // a HashMap in the CallbackManager instance, sCallbackManager.
+        // Because FacebookProvider which contains an instance of an activity, mCallbackObject,
+        // is contained in sCallbackManager, that activity will not be garbage collected.
+        // Thus, we have leaked an Activity.
```

Comment:
Do we want to keep this explanation? It's a bit lengthy and could probably be replaced with `// Fixes activity leaks`.

Here is the Facebook code in question:
```java
CallbackManager.Factory.create()

public static CallbackManager create() {
    return new CallbackManagerImpl();
}
```

```java
loginManager.registerCallback(sCallbackManager, ***this***);
```

->

```java
/**
 * Registers a login callback to the given callback manager.
 * @param callbackManager The callback manager that will encapsulate the callback.
 * @param callback The login callback that will be called on login completion.
 */
public void registerCallback(
final CallbackManager callbackManager,
final FacebookCallback<LoginResult> callback) {
    if (!(callbackManager instanceof CallbackManagerImpl)) {
        throw new FacebookException("Unexpected CallbackManager, " +
                                            "please use the provided Factory.");
    }
    ((CallbackManagerImpl) callbackManager).registerCallback( // sCallbackManager is storing callback into itself
            CallbackManagerImpl.RequestCodeOffset.Login.toRequestCode(),
            new CallbackManagerImpl.Callback() {
                @Override
                public boolean onActivityResult(int resultCode, Intent data) {
                    return LoginManager.this.onActivityResult(
                            resultCode,
                            data,
                            ***callback***);
                }
            }
    );
}
```

->

```java
private Map<Integer, CallbackManagerImpl.Callback> callbacks = new HashMap<>();

public void registerCallback(int requestCode, Callback callback) {
    Validate.notNull(callback, "callback");
    callbacks.put(requestCode, ***callback***); // Callback is now in callbacks HashMap
}
```